File: src/ca/joelathiessen/kaly2/tests/pc/RobotTest.java
package ca.joelathiessen.kaly2.tests.pc;

import static org.junit.Assert.*;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.mockito.Mockito;

import ca.joelathiessen.kaly2.main.Robot;
import ca.joelathiessen.kaly2.main.subconscious.Measurement;
import ca.joelathiessen.kaly2.main.subconscious.Subconscious;

public class RobotTest {

  Robot robot;

  @Before
  public void setUp() {
    Subconscious sub = Mockito.mock(Subconscious.class);
    ConcurrentLinkedQueue<ArrayList<Measurement>> sweeps =
        new ConcurrentLinkedQueue<ArrayList<Measurement>>();
    robot = new Robot(sub, sweeps);
  }

  @Test
  public void testRobot() {
    assertNotNull(robot); // is this sort of test actually necessary?
  }
  
  @Test
  public void testConstructRobot_PassedValuesUnmodified() {

    Subconscious sub = Mockito.mock(Subconscious.class);
    ConcurrentLinkedQueue<ArrayList<Measurement>> sweeps =
        new ConcurrentLinkedQueue<ArrayList<Measurement>>();
    
    Subconscious subCopy = sub;
    ConcurrentLinkedQueue<ArrayList<Measurement>> sweepsCopy = sweeps;
    
    Robot robot2 = new Robot(sub, sweeps);
    robot2.toString();
    
    assertEquals(sub,subCopy);
    assertEquals(sweeps,sweepsCopy);
  }

  @Test
  public void testStartStopRobotSimple() {
    robot.startRobot();
    assertTrue(robot.isRunning());
    robot.stopRobot();
    assertFalse(robot.isRunning());
  }

  @Test
  public void testStartStopRobotRunning() {
    Thread thread = new Thread(robot);
    thread.start();
    robot.startRobot();
    assertTrue(robot.isRunning());
    robot.stopRobot();
    assertFalse(robot.isRunning());

    thread.interrupt();
    try {
      thread.join(150);
    } catch (InterruptedException e) {
      fail("Robot thread was interupted but didn't stop in a timely fashion");
    }
    assertFalse(thread.isAlive());
  }

}

File: src/test/kotlin/ca/joelathiessen/kaly2/tests/pc/unit/RobotTest.kt
package ca.joelathiessen.kaly2.tests.pc.unit

import ca.joelathiessen.kaly2.Measurement
import ca.joelathiessen.kaly2.Robot
import ca.joelathiessen.kaly2.subconscious.Subconscious
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.powermock.core.classloader.annotations.PrepareForTest
import org.powermock.modules.junit4.PowerMockRunner
import java.util.ArrayList
import java.util.concurrent.ConcurrentLinkedQueue

@RunWith(PowerMockRunner::class)
@PrepareForTest(Subconscious::class, Measurement::class)
class RobotTest {

    internal lateinit var robot: Robot

    @Before
    fun setUp() {
        val sub = Mockito.mock(Subconscious::class.java)
        val sweeps = ConcurrentLinkedQueue<ArrayList<Measurement>>()
        robot = Robot(sub, sweeps)
    }

    @Test
    fun testRobot() {
        assertNotNull(robot) // is this sort of test actually necessary?
    }

    @Test
    fun testConstructRobot_PassedValuesUnmodified() {
        val sub = Mockito.mock(Subconscious::class.java)
        val sweeps = ConcurrentLinkedQueue<ArrayList<Measurement>>()

        val subCopy = sub
        val sweepsCopy = sweeps

        val robot2 = Robot(sub, sweeps)
        robot2.toString()

        assertEquals(sub, subCopy)
        assertEquals(sweeps, sweepsCopy)
    }

    @Test
    fun testStartStopRobotSimple() {
        robot.startRobot()

        assertTrue(robot.isRunning)

        robot.stopRobot()

        assertFalse(robot.isRunning)
    }

    @Test
    fun testStartStopRobotRunning() {
        val thread = Thread(robot)
        thread.start()
        robot.startRobot()

        assertTrue(robot.isRunning)

        robot.stopRobot()

        assertFalse(robot.isRunning)

        thread.interrupt()

        try {
            thread.join(150)
        } catch (e: InterruptedException) {
            fail("Robot thread was interupted but didn't stop in a timely fashion")
        }

    }

}
